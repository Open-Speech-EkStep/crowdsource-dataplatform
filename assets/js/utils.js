const { HOUR_IN_SECONDS, SIXTY } = require("./constants");

const ALL_LANGUAGES = [
    {value: "Assamese",id: "as", text: "অসমীয়া", hasLocaleText: true},
    {value: "Bengali", id: "bn", text: "বাংলা", hasLocaleText: true},
    {value: "English", id: "en", text: "English", hasLocaleText: true},
    {value: "Gujarati", id: "gu", text: "ગુજરાતી", hasLocaleText: true},
    {value: "Hindi", id: "hi", text: "हिंदी", hasLocaleText: true},
    {value: "Kannada", id: "kn", text: "ಕನ್ನಡ", hasLocaleText: true},
    {value: "Malayalam", id: "ml", text: "മലയാളം", hasLocaleText: true},
    {value: "Marathi", id: "mr", text: "मराठी", hasLocaleText: true},
    {value: "Odia", id: "or", text: "ଓଡିଆ", hasLocaleText: true},
    {value: "Punjabi", id: "pa", text: "ਪੰਜਾਬੀ", hasLocaleText: true},
    {value: "Tamil", id: "ta", text: "தமிழ்", hasLocaleText: true},
    {value: "Telugu", id: "te", text: "తెలుగు", hasLocaleText: true}];

function convertPXToVH(px) {
    return px * (100 / document.documentElement.clientHeight);
}

function setPageContentHeight() {
    const $footer = $('footer');
    const $nav = $('.navbar');
    const edgeHeightInPixel = $footer.outerHeight() + $nav.outerHeight()
    const contentHeightInVH = 100 - convertPXToVH(edgeHeightInPixel)
    $('#content-wrapper').css('min-height', contentHeightInVH + 'vh');
}

function toggleFooterPosition() {
    const $footer = $('footer');
    $footer.toggleClass('fixed-bottom')
    $footer.toggleClass('bottom')
}

function fetchLocationInfo() {
    //https://api.ipify.org/?format=json
    return fetch('https://www.cloudflare.com/cdn-cgi/trace').then(res => res.text()).then(ipAddressText => {
        const dataArray = ipAddressText.split('\n');
        let ipAddress = "";
        for (let ind in dataArray) {
            if (dataArray[ind].startsWith("ip=")) {
                ipAddress = dataArray[ind].replace('ip=', '');
                break;
            }
        }
        if (ipAddress.length !== 0) {
            return fetch(`/location-info?ip=${ipAddress}`);
        } else {
            return new Promise((resolve, reject) => {
                reject("Ip Address not available")
            })
        }
    });
}

const performAPIRequest = (url) => {
    return fetch(url).then((data) => {
        if(!data.ok) {
            throw Error(data.statusText || 'HTTP error');
        } else {
            return Promise.resolve(data.json());
        }
    });
}

const getLocaleString = function() {
    performAPIRequest('/get-locale-strings')
        .then((response) => {
            localStorage.setItem('localeString', JSON.stringify(response));
        });
}

const updateLocaleLanguagesDropdown = (language) => {
    const dropDown = $('#localisation_dropdown');
    const localeLang = ALL_LANGUAGES.find(ele => ele.value === language);
    if (language.toLowerCase() === "english" || localeLang.hasLocaleText === false) {
        dropDown.html('<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>');
    } else {
        dropDown.html(`<a id="english" class="dropdown-item" href="/changeLocale/en">English</a>
        <a id=${localeLang.value} class="dropdown-item" href="/changeLocale/${localeLang.id}">${localeLang.text}</a>`);
    }
}

const calculateTime = function (totalSeconds, isSeconds = true) {
    const hours = Math.floor(totalSeconds / HOUR_IN_SECONDS);
    const remainingAfterHours = totalSeconds % HOUR_IN_SECONDS;
    const minutes = Math.floor(remainingAfterHours / SIXTY);
    const seconds = Math.round(remainingAfterHours % SIXTY);
    if (isSeconds) {
        return { hours, minutes, seconds };
    } else {
        return { hours, minutes };
    }
};

const formatTime = function (hours, minutes = 0, seconds = 0) {
    let result = "";
    if (hours > 0) {
        result += `${hours} hrs `;
    }
    if (minutes > 0) {
        result += `${minutes} min `;
    }
    if (hours === 0 && minutes === 0 && seconds > 0) {
        result += `${seconds} sec `;
    }
    return result.substr(0, result.length - 1);
};

module.exports = { setPageContentHeight, toggleFooterPosition, fetchLocationInfo, updateLocaleLanguagesDropdown ,calculateTime, formatTime, getLocaleString, performAPIRequest}
