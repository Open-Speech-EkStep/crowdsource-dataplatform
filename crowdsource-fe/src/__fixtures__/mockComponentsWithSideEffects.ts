jest.mock('components/ContributionStats', () => () => 'ContributionStats');
jest.mock('components/Feedback', () => () => 'Feedback');
jest.mock('components/Report', () => () => 'Report');
jest.mock('components/TargetProgress', () => () => 'Feedback');
jest.mock('components/ContributionTracker', () => () => 'ContributionTracker');
jest.mock('components/Charts/BarChart', () => () => 'BarChart');
jest.mock('components/DataLastUpdated', () => () => 'DataLastUpdated');

export {};
