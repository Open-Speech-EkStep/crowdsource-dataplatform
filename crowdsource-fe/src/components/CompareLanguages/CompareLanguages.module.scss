.heading {
  border-bottom: 1px dashed var(--primary-40-color);

  @include media-breakpoint-up(md) {
    border-bottom: 0;
    border-right: 1px dashed var(--primary-40-color);
  }
}

.chart {
  height: 100%;
}

.chartContainer {
  height: 100%;
}
