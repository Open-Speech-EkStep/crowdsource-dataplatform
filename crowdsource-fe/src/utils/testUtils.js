import { render } from '@testing-library/react';

const Wrapper = ({ children }) => children;

const customRender = (ui, options) => render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };
