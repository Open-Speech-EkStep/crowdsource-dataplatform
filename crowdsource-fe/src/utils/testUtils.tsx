/* eslint-disable jest/no-export */
import type { ReactNode, ReactElement, ComponentType } from 'react';

import type { RenderOptions, RenderResult } from '@testing-library/react';
import { render, act } from '@testing-library/react';
import { axe } from 'jest-axe';
import puppeteer from 'puppeteer';
import type { Browser } from 'puppeteer';

const Wrapper: ComponentType = ({ children }: { children?: ReactNode }) => children as ReactElement;

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: Wrapper, ...options });

// re-export everything
export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { renderHook } from '@testing-library/react-hooks';

// override render method
export { customRender as render };

export const verifyAxeTest = ({ container }: { container: RenderResult['container'] }) => {
  it('should not fail an axe audit', async () => {
    let axeResult;

    await act(async () => {
      axeResult = await axe(container);
    });

    expect(axeResult).toHaveNoViolations();
  });
};

export const verifyVRTest = (title: string, url: string) => {
  describe(`${title} VR test`, () => {
    let browser: Browser;

    beforeAll(async () => {
      browser = await puppeteer.launch({
        headless: true,
        devtools: false,
        ignoreHTTPSErrors: true,
        args: ['--start-fullscreen'],
      });
    });

    it('match snapshot', async () => {
      const page = await browser.newPage();
      await page.setViewport({ width: 1920, height: 1080 });

      await page.goto(url, {
        waitUntil: 'networkidle2',
      });

      await page.addStyleTag({
        content: `
      *,
      *::after,
      *::before {
          transition-delay: 0s !important;
          transition-duration: 0s !important;
          animation-delay: -0.0001s !important;
          animation-duration: 0s !important;
          animation-play-state: paused !important;
          caret-color: transparent !important;
      }
      `,
      });

      const image = await page.screenshot({
        fullPage: true,
      });

      expect(image).toMatchImageSnapshot();
    });

    afterAll(async () => {
      await browser.close();
    });
  });
};
