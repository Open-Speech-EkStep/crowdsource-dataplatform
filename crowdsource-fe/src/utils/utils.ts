import { jsPDF } from 'jspdf';
import { i18n } from 'next-i18next';

import { ErrorStatusCode } from 'constants/errorStatusCode';
import { INITIATIVES_MAPPING } from 'constants/initiativeConstants';
import { KEYBOARD_ERROR, LANGUAGE_UNICODE, OTHER_LANGUAGE_UNICODE } from 'constants/Keyboard';
import type { InitiativeType } from 'types/InitiativeType';

import apiPaths from '../constants/apiPaths';

var platform = require('platform');

export const convertIntoHrsFormat = (data: any, isSeconds = true) => {
  const hours = Math.floor(data / 3600);
  const remainingAfterHours = data % 3600;
  const minutes = Math.floor(remainingAfterHours / 60);
  const seconds = Math.round(remainingAfterHours % 60);
  if (isSeconds) {
    return { hours, minutes, seconds };
  } else {
    return { hours, minutes };
  }
};

export const roundOffValue = (average: any, precision: number) => {
  const formattedAverage = average.toFixed(precision);
  return formattedAverage;
};

export const capitalizeFirstLetter = (strValue: string) => {
  return strValue?.charAt(0).toUpperCase() + strValue?.slice(1);
};

const translateText = (text: string) => i18n?.t(text);

export const formatTime = (hours: number, minutes = 0, seconds = 0) => {
  const hrStr = translateText('hours1');
  const minStr = translateText('minutes1');
  const secStr = translateText('seconds1');
  let result = '';
  if (hours > 0) {
    result += `${hours} ${hrStr} `;
  }
  if (minutes > 0) {
    result += `${minutes} ${minStr} `;
  }
  if (hours === 0 && minutes === 0 && seconds > 0) {
    result += `${seconds} ${secStr} `;
  }

  if (hours === 0 && minutes === 0 && seconds === 0) {
    result += `0 ${secStr} `;
  }

  return result.substr(0, result.length - 1);
};

export const convertTimeFormat = (value: any) => {
  const { hours, minutes, seconds } = convertIntoHrsFormat(Number(value) * 60 * 60);
  const data = formatTime(hours, minutes, seconds);
  return data;
};

const roundDuration = (value: number) => {
  return Math.round(value * 100) / 100;
};

export const getHoursValue = (value: any) => {
  return roundDuration(value);
};

export const getMinutesValue = (value: any) => {
  return roundDuration(value * 60);
};

export const getHoursText = (value: any) => {
  return `${getHoursValue(value)} ${i18n?.t('hours2')}`;
};

export const getMinutesText = (value: any) => {
  return `${getMinutesValue(value)} ${i18n?.t('minutes2')}`;
};

export const isSunoOrBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['asr', 'text'];
  return boloOrSuno.includes(value);
};

export const isBoloInitiative = (value: InitiativeType) => {
  const boloOrSuno = ['text'];
  return boloOrSuno.includes(value);
};

export const findInputError = (text: string, currentModule: string, language: string) => {
  if (!text.length) return;
  const newText = text.replace(/\s/g, ''); //read input value, and remove "space" by replace \s
  let error: any = KEYBOARD_ERROR.language;
  const isSunoModule = currentModule === INITIATIVES_MAPPING.suno;
  const specialSymbols = /[\u0021-\u002F\u003A-\u0040\u005B-\u0060\u007B-\u007E\u0964-\u0965]/;
  if (currentModule === INITIATIVES_MAPPING.suno && specialSymbols.test(newText) === true) {
    return KEYBOARD_ERROR.symbol;
  }
  Object.entries(isSunoModule ? LANGUAGE_UNICODE : OTHER_LANGUAGE_UNICODE).forEach(([key, value]) => {
    if (value.test(newText)) {
      //Check Unicode to see which one is true
      if (key.toLowerCase() === language.toLowerCase()) {
        error = KEYBOARD_ERROR.noError;
      }
    }
  });
  return error;
};

export const fetchLocationInfo = async () => {
  const result = await fetch('https://www.cloudflare.com/cdn-cgi/trace')
    .then(res => res.text())
    .then(async ipAddressText => {
      const dataArray = ipAddressText.split('\n');
      let ipAddress = '';
      for (let ind in dataArray) {
        if (dataArray[ind].startsWith('ip=')) {
          ipAddress = dataArray[ind].replace('ip=', '');
          break;
        }
      }
      if (ipAddress.length !== 0) {
        const data = await fetch(`${apiPaths.locationInfo}?ip=${ipAddress}`);
        return data && data.json();
      } else {
        return new Promise((resolve, reject) => {
          reject('Ip Address not available');
        });
      }
    });

  return result;
};

export const getDeviceInfo = () => {
  const os = platform.os;
  return os.family + ' ' + os.version;
};

export const getBrowserInfo = () => {
  return platform.name + ' ' + platform.version;
};

export const downloadBadge = (
  locale: string | undefined,
  initiative: string,
  source: string,
  badgeType: string,
  badgeId: string
) => {
  const pdf = new jsPDF();
  const img = new Image();
  img.src = `/images/${locale}/badges/${locale}_${initiative}_${badgeType}_${source}.png`;
  img.crossOrigin = 'Anonymous';
  img.onload = function () {
    pdf.addImage(img, 50, 10, 105, 130);
    pdf.save(`${badgeType}-badge.pdf`);
  };

  pdf.text(`Badge Id : ${badgeId}`, 36, 190);
};

export const groupBy = function (list: any, keyName: any) {
  if (!list || list.length === 0 || !list[0][keyName]) {
    return [];
  }
  return list.reduce(function (pair: any, key: any) {
    (pair[key[keyName]] = pair[key[keyName]] || []).push(key);
    return pair;
  }, {});
};

/* istanbul ignore next */
export const visualize = (visualizer: any, analyser: any) => {
  const canvasCtx = visualizer.getContext('2d');
  const bufferLength = analyser.frequencyBinCount;
  const dataArray = new Uint8Array(bufferLength);
  const WIDTH = visualizer.width;
  const HEIGHT = visualizer.height;
  // TODO do we need to limit the number of time visualize refreshes per second
  // so that it can run on Android processors without causing audio to drop?
  function draw() {
    // this is more efficient than calling with processor.onaudioprocess
    // and sending floatarray with each call...
    requestAnimationFrame(draw);
    analyser.getByteTimeDomainData(dataArray);
    // canvasCtx.fillStyle = 'rgb(255, 255, 255, 0)';
    // canvasCtx.fillRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.clearRect(0, 0, WIDTH, HEIGHT);
    canvasCtx.lineWidth = 2;
    canvasCtx.strokeStyle = 'rgb(0,123,255)';
    canvasCtx.beginPath();
    const sliceWidth = (WIDTH * 1.0) / bufferLength;
    let x = 0;
    for (let i = 0; i < bufferLength; i++) {
      let v = dataArray[i] / 128.0; // uint8
      let y = (v * HEIGHT) / 2; // uint8
      if (i === 0) {
        canvasCtx.moveTo(x, y);
      } else {
        canvasCtx.lineTo(x, y);
      }
      x += sliceWidth;
    }
    canvasCtx.lineTo(visualizer.width, visualizer.height / 2);
    canvasCtx.stroke();
  }

  draw();
};

export const getErrorMsg = (error: any) => {
  if (error && error.status === ErrorStatusCode.TOO_MANY_REQUEST) {
    return 'multipleRequestApiError';
  } else {
    return 'apiFailureError';
  }
};
