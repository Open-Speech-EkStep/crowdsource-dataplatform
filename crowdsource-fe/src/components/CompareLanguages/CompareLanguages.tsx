import React, { useMemo, useEffect } from 'react';

import { useTranslation, i18n } from 'next-i18next';

import { BarChart } from 'components/Charts';
import apiPaths from 'constants/apiPaths';
import { INITIATIVES_MEDIA, INITIATIVES_MEDIA_MAPPING } from 'constants/initiativeConstants';
import localStorageConstants from 'constants/localStorageConstants';
import useFetch from 'hooks/useFetch';
import useLocalStorage from 'hooks/useLocalStorage';
import type { CumulativeDataByLanguage } from 'types/CumulativeCount';
import type { Initiative } from 'types/Initiatives';
import type { InitiativeType } from 'types/InitiativeType';
import type { TopLanguagesByHours } from 'types/TopLanguages';
import { convertTimeFormat, isBoloInitiative } from 'utils/utils';

import styles from './CompareLanguages.module.scss';

const mapChartData = (
  data: Array<CumulativeDataByLanguage> | undefined,
  key: string,
  langauge: string,
  initiativeMedia: InitiativeType,
  value: string
) => {
  if (langauge && data && data.length) {
    const contributedLanguageHours = data.find(
      (item: CumulativeDataByLanguage) => item.language === langauge && item.type === initiativeMedia
    );
    const topLanguageArray = [];
    if (contributedLanguageHours) {
      topLanguageArray.push(contributedLanguageHours);
    } else {
      topLanguageArray.push({
        language: langauge,
        [value]: 0.0,
      });
    }
    let remainingLanguage = data.filter((item: CumulativeDataByLanguage) => item?.language !== langauge);
    remainingLanguage = remainingLanguage.sort((a: any, b: any) =>
      Number(a[key]) > Number(b[key]) ? -1 : 1
    );
    return topLanguageArray.concat(remainingLanguage.slice(0, 3));
  }
};

const translateCategory = (language: string) => {
  return language
    .split('-')
    .map(language => i18n?.t(language.toLowerCase()))
    .join('-');
};

const getTopLanguagesByHoursChartData = (
  contributionValue: string,
  initiativeType: InitiativeType,
  key: string,
  topLanguagesByHours?: TopLanguagesByHours[]
) => {
  return (
    topLanguagesByHours?.map((topLanguageByHours: any) => ({
      category: translateCategory(topLanguageByHours?.language),
      value: (topLanguageByHours as any)[contributionValue],
      tooltipText: isBoloInitiative(initiativeType)
        ? convertTimeFormat(topLanguageByHours?.total_contributions)
        : `${topLanguageByHours[key]} ${i18n?.t(`${initiativeType}BarGraphTooltip`)}`,
    })) ?? []
  );
};

interface CompareLanguagesProps {
  initiative: Initiative;
  dataBindigValue: string;
  graphLabel: string;
  isTopLanguage: any;
  graphHeading: string;
  showHeader: boolean;
}

const CompareLanguages = (props: CompareLanguagesProps) => {
  const { t } = useTranslation();
  const [contributionLanguage] = useLocalStorage<string>(localStorageConstants.contributionLanguage);
  const [translatedLanguage] = useLocalStorage<string>(localStorageConstants.translatedLanguage);
  const { data: cumulativeDataByLanguage, mutate: hrsMutate } = useFetch<Array<CumulativeDataByLanguage>>(
    apiPaths.cumulativeDataByLanguage,
    { revalidateOnMount: false }
  );

  const translatedContributionLanguage = `${t(`${contributionLanguage?.toLowerCase()}`)}${
    translatedLanguage ? `-${t(`${translatedLanguage?.toLowerCase()}`)}` : ''
  }`;

  useEffect(() => {
    if (contributionLanguage) {
      hrsMutate();
    }
  }, [contributionLanguage, translatedLanguage, hrsMutate]);

  let cumulativeData: Array<CumulativeDataByLanguage> | undefined;

  cumulativeData = cumulativeDataByLanguage?.filter(
    topLanguageByHours => topLanguageByHours.type === INITIATIVES_MEDIA_MAPPING[props.initiative]
  );

  const language =
    INITIATIVES_MEDIA_MAPPING[props.initiative] === INITIATIVES_MEDIA.parallel
      ? `${contributionLanguage}-${translatedLanguage}`
      : contributionLanguage;

  const topLanguageHrsData = mapChartData(
    cumulativeData,
    props.dataBindigValue,
    language as string,
    INITIATIVES_MEDIA_MAPPING[props.initiative],
    props.dataBindigValue
  );

  const chartLegendDetails = useMemo(
    () => ({
      colors: ['#F7CC56', '#F7CC56', '#F7CC56', '#EF8537'],
      xAxisLabel: t('month'),
    }),
    [t]
  );

  useEffect(() => {
    const sortingLanguages =
      topLanguageHrsData &&
      topLanguageHrsData.sort((a: any, b: any) =>
        Number(a[props.dataBindigValue]) > Number(b[props.dataBindigValue]) ? -1 : 1
      );
    const top3language = sortingLanguages && sortingLanguages.slice(0, 3);
    const found = top3language?.some(el => el.language === contributionLanguage);
    const isLanguageTop = found;
    props.isTopLanguage(isLanguageTop ? 'keep' : 'see');
  }, [contributionLanguage, props, topLanguageHrsData]);

  const barChartData = useMemo(
    () => ({
      data: getTopLanguagesByHoursChartData(
        props.dataBindigValue,
        INITIATIVES_MEDIA_MAPPING[props.initiative],
        props.dataBindigValue,
        topLanguageHrsData
      ),
      yAxisLabel: props.graphLabel,
      ...chartLegendDetails,
    }),
    [chartLegendDetails, props.dataBindigValue, props.graphLabel, props.initiative, topLanguageHrsData]
  );

  return barChartData.data.length ? (
    <div data-testid="CompareLanguages" className={styles.chartContainer}>
      {props.showHeader && <h5 className="fw-light mb-3 text-center">{props.graphHeading}</h5>}
      <div className={styles.chart}>
        <BarChart id="tybar_chart" data={barChartData} />
      </div>
      <span className="display-6 d-block text-center text-primary-60">
        {t(`${props.initiative}contributionVsTopLanguage`, { language: translatedContributionLanguage })}
      </span>
    </div>
  ) : null;
};

export default CompareLanguages;
