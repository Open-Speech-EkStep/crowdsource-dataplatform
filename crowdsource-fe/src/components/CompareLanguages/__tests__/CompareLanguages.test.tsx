import '__fixtures__/mockComponentsWithSideEffects';

import { when } from 'jest-when';

import { render, screen, waitFor } from 'utils/testUtils';

import CompareLanguages from '../CompareLanguages';

describe('CompareLanguages', () => {
  const setup = async (initiative: 'suno' | 'likho' | 'bolo' | 'dekho', source: string) => {
    const language = 'Hindi';

    when(localStorage.getItem)
      .calledWith('contributionLanguage')
      .mockImplementation(() => language);

    when(localStorage.getItem)
      .calledWith('likho_to-language')
      .mockImplementation(() => 'English');

    fetchMock.doMockOnceIf('/aggregated-json/cumulativeDataByLanguage.json').mockResponseOnce(
      JSON.stringify([
        {
          language: 'English',
          total_contribution_count: 37,
          total_contributions: 0.058,
          total_speakers: 10,
          total_validation_count: 2,
          total_validations: 0.001,
          type: 'asr',
        },
        {
          language: 'English',
          total_contribution_count: 37,
          total_contributions: 0.058,
          total_speakers: 10,
          total_validation_count: 2,
          total_validations: 0.001,
          type: 'ocr',
        },
        {
          language: 'Hindi',
          total_contribution_count: 37,
          total_contributions: 0.058,
          total_speakers: 10,
          total_validation_count: 2,
          total_validations: 0.001,
          type: 'text',
        },
        {
          language: 'Hindi',
          total_contribution_count: 37,
          total_contributions: 0.068,
          total_speakers: 10,
          total_validation_count: 2,
          total_validations: 0.001,
          type: 'asr',
        },
        {
          language: 'Hindi-English',
          total_contribution_count: 37,
          total_contributions: 0.058,
          total_speakers: 10,
          total_validation_count: 2,
          total_validations: 0.001,
          type: 'parallel',
        },
        {
          language: 'Kannada',
          total_contribution_count: 37,
          total_contributions: 0.054,
          total_speakers: 10,
          total_validation_count: 2,
          total_validations: 0.001,
          type: 'asr',
        },
      ])
    );

    const renderResult = render(
      <CompareLanguages
        initiative={initiative}
        source={source}
        dataBindigValue="total_contributions"
        graphLabel="totalSentences"
      />
    );
    await waitFor(() => {
      expect(localStorage.getItem).toBeCalled();
    });
    await waitFor(() => {
      expect(fetchMock).toBeCalledWith('/aggregated-json/cumulativeDataByLanguage.json');
    });
    return renderResult;
  };

  it('should render the component and matches it against stored snapshot', async () => {
    const { asFragment } = await setup('suno', 'contribute');

    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the chart component after api gives data', async () => {
    await setup('likho', 'validate');

    expect(screen.getByTestId('CompareLanguages').children.length).toBe(1);
  });
});
