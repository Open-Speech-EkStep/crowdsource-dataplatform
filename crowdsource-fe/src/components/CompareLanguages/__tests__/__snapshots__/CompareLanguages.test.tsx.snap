// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompareLanguages should render the component and matches it against stored snapshot 1`] = `
<DocumentFragment>
  <div
    class="chartContainer"
    data-testid="CompareLanguages"
  >
    <h5
      class="fw-light mb-3 text-center"
    >
      Participate to keep Bengali in top 3
    </h5>
    <div
      class="chart"
    >
      BarChart
    </div>
    <span
      class="display-6 d-block text-center text-primary-60"
    >
      sunocontributionVsTopLanguage
    </span>
  </div>
</DocumentFragment>
`;
